src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -112,17 +113,20 @@ public void Shutdown()
             {
                 var handleActiveFile = true;
                 using var _ = PooledHashSet<DocumentId>.GetInstance(out var documentSet);
+                using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
 
                 foreach (var stateSet in stateSets)
                 {
                     var projectIds = stateSet.GetProjectsWithDiagnostics();
                     foreach (var projectId in projectIds)
                     {
                         stateSet.CollectDocumentsWithDiagnostics(projectId, documentSet);
-                        RaiseProjectDiagnosticsRemoved(stateSet, projectId, documentSet, handleActiveFile, raiseEvents);
+                        AddProjectDiagnosticsRemovedArgs(ref argsBuilder.AsRef(), stateSet, projectId, documentSet, handleActiveFile);
                         documentSet.Clear();
                     }
                 }
+
+                raiseEvents(argsBuilder.ToImmutableAndClear());
```

Comment:
>argsBuilder [](http://example.com/codeflow?start=28&length=11)

check count before calling raiseEvents?

Reply:
➡️ `raiseEvents` checks the count itself, and `ToImmutableAndClear()` is a NOP for empty builder.